onnx/onnx.proto

```diff
@@ -64,21 +65,34 @@ message AttributeProto {
   repeated GraphProto graphs = 11;    // list of graph
 }
 
+// Defines argument information set, except argument name.
+// The argument could be an input or an output of a graph or node.
+message ArgInfoProto {
+  // This field MUST be present in this version of the IR.  
+  optional TypeProto type = 1;
+  optional TensorShapeProto shape = 2;
+}
+
 // NodeProto stores a node that is similar to the notion of "layer"
 // or "operator" in many deep learning frameworks. For example, it can be a
 // node of type "Conv" that takes in an image, a filter tensor and a bias
 // tensor, and produces the convolved output.
 message NodeProto {
-  repeated string input = 1;   // namespace Tensor
-  repeated string output = 2;  // namespace Tensor
-  optional string name = 3;    // namespace Node
-  optional string op_type = 4; // namespace Operator
+  repeated string input = 1;    // namespace Value
+  repeated string output = 2;   // namespace Value
+  optional string name = 3;     // namespace Node
+  optional string op_type = 4;  // namespace Operator
 
   // Additional named attributes.
   repeated AttributeProto attribute = 5;
 
   // A human-readable documentation for this graph. Markdown is allowed.
   optional string doc_string = 6;
+
+  // Information including type and shape for each input and output of
```

Comment:
I'm a little confused what is the role of specification at the individual Node level. In particular all inputs of the node are some other Node's outputs, so the type/shape information is redundant. If we wanted to preserve types and shapes of intermediate value it feel more natural to just record them in "value_name -> ArgInfoProto" directly in the graph.

Second, how does it connect with the type specifications and inferences #55 ? If we're moving towards full automatic shape inference there is little need to record the shapes of the intermediate values in the model proto.

Reply:
On your first point, you are correct that the type declarations on Node require redundant specification on both the input and output side. If we keep this rep, we need to define the rules for what happens when they TypeProto is not identical.   I'll defer to Yuan as to whether encoding a single table that declares the type of each edge/named-value is better or worse.

On your second point, one could take a graph proto and the library protos for all referenced operators and either (a) fill in absent Node type decls or (b) check the ones that are there.  Note that in Yuan's PR, he's only requiring type decls on Graph input/output, not Node (which as you state can be inferred).

Reply:
Agree with Don. On the second point, this allows the possibility of a type-inference being a stand-alone tool, that takes a graph-proto as input and writes a (fully type-annotated) graph-proto as output.

Reply:
On the second point - what is the use case of recording intermediate tensor shapes in the serialized model? I'm a bit biased towards keeping the minimum necessary functionality for model interchange in ONNX whereas inference of shapes can be handled by corresponding frameworks. As you said, we can always infer shapes based on Node definitions, operator specs and graph inputs.

Reply:
On the first point, I am fine with encoding them as "value_name -> ArgInfoProto" in graph. 

On the second point, I don't think that this PR and Don's PR commit to fully automatic shape inference. We can discuss if that is the direction we should go and the issues we might run into.
For example, do we really want to require all participating frameworks to perform fully automatic shape inference? Even for type inference, while we will definitely d it, it is uncertain to me if we want to require it for all frameworks.

Reply:
For type/shape inference, is it possible to do it in the onnx level so that participating frameworks don't need to implement it again?

Reply:
We can do something like:
1. provide a set of default type/shape inference functions
2. offer a way to override the default inference rules.

Reply:
Dmytro, we should definitely collapse the GraphProto.input/GraphProto.input_arg_info fields for both compactness and to make it less likely that folks don't screw it up..

I suspect once we write down the type inference rules, someone will contribute a reference implementation. I agree that would be pure goodness and will help us on interop.